addons/account/models/account_move.py

```diff
@@ -4703,277 +4703,336 @@ def _mail_track(self, tracked_fields, initial):
+            # currency of the opposite line.
+            if debit_vals['currency'] == company_currency:
+                partial_debit_amount_currency = partial_amount
+            else:
+                partial_debit_amount_currency = min_recon_amount
+            if credit_vals['currency'] == company_currency:
+                partial_credit_amount_currency = partial_amount
+            else:
+                partial_credit_amount_currency = min_recon_amount
 
-            # ==== Computation of partial amounts ====
+        # Computation of the partial exchange difference. You can skip this part using the
+        # `no_exchange_difference` context key (when reconciling an exchange difference for example).
+        if not self._context.get('no_exchange_difference'):
+            exchange_lines_to_fix = self.env['account.move.line']
+            amounts_list = []
             if recon_currency == company_currency:
-                # Compute the partial amount expressed in company currency.
-                partial_amount = min_recon_amount
-
-                # Compute the partial amount expressed in foreign currency.
-                if debit_rate:
-                    partial_debit_amount_currency = debit_line.currency_id.round(debit_rate * min_recon_amount)
-                    partial_debit_amount_currency = min(partial_debit_amount_currency, remaining_debit_amount_curr)
-                else:
-                    partial_debit_amount_currency = 0.0
-                if credit_rate:
-                    partial_credit_amount_currency = credit_line.currency_id.round(credit_rate * min_recon_amount)
-                    partial_credit_amount_currency = min(partial_credit_amount_currency, -remaining_credit_amount_curr)
-                else:
-                    partial_credit_amount_currency = 0.0
+                if debit_fully_matched:
+                    debit_exchange_amount = remaining_debit_amount_curr - partial_debit_amount_currency
+                    if not debit_vals['currency'].is_zero(debit_exchange_amount):
+                        if debit_vals.get('record'):
+                            exchange_lines_to_fix += debit_vals['record']
+                        amounts_list.append({'amount_residual_currency': debit_exchange_amount})
+                        remaining_debit_amount_curr -= debit_exchange_amount
+                if credit_fully_matched:
+                    credit_exchange_amount = remaining_credit_amount_curr + partial_credit_amount_currency
+                    if not credit_vals['currency'].is_zero(credit_exchange_amount):
+                        if credit_vals.get('record'):
+                            exchange_lines_to_fix += credit_vals['record']
+                        amounts_list.append({'amount_residual_currency': credit_exchange_amount})
+                        remaining_credit_amount_curr += credit_exchange_amount
```

Comment:
I think quite the same than above can be done to make it simpler to read/understand